README.md
w# Keboola GoodData Writer

## Tables for Shared Config
- columns in bold should be primary keys
- columns in italics should be indices

### jobs
**id**,runId,projectId,writerId,token,tokenId,tokenDesc,tokenOwnerName,initializedBy,createdTime,startTime,endTime,command,pid,dataset,xmlFile,parameters,result,gdWriteStartTime,gdWriteBytes,status,log,_projectIdWriterId,batchId,queueId_

### projects
**pid**,projectId,writerId,backendUrl,accessToken,createdTime,_projectIdWriterId_
 
### users
**uid**,email,projectId,writerId,createdTime,_projectIdWriterId_
 
### projects_to_delete
**pid**,projectId,writerId,dev,createdTime,_deletedTime_

### users_to_delete
**uid**,projectId,writerId,email,dev,createdTime,_deletedTime_

CHANGELOG.md
## 2014-05-15 (v 5.14)
- creating of writer with existing GD project

## 2014-05-12 (v 5.13)
- optimization of SLI hashes

## 2014-03-25 (v 5.12)
- async export from Storage API

## 2014-04-24 (v 5.11)
- update to Syrup 1.3

## 2014-03-17 (v 5.10)
- refactorings

## 2014-03-04 (v 5.9)
- added POST call `reset-table` which deletes dataset in GoodData project belonging to the table

## 2014-02-26 (v 5.8)
- added POST call `reset-project` which deletes GoodData project and creates new one

## 2014-02-18 (v 5.7)
- attribute `maintenance` for postponing of Writer's jobs process

## 2014-02-11 (v 5.6)
- uploading of date dimensions separated as single job
- added POST API call `upload-date-dimension`

## 2014-01-31 (v 5.5)
- added parameters `status` and `tokenId` to filter jobs list (API call `jobs`)

## 2014-01-28 (v 5.4)
- added parameters `command` and `tableId` to filter jobs list (API call `jobs`)

## 2014-01-27 (v 5.3)
- migration of maql execution in GD to polling (`/ldm/manage2`)

## 2014-01-16 (v 5.2)
- maql generated by CL tool is executed directly via Rest API

## 2014-01-15 (v 5.1)
- direct upload of csv to WebDav without saving to our server's disk

## 2014-01-13 (v 5.0)
- changed configuration of datasets

## 2013-11-29 (v 4.13)
- added POST API call `proxy` which can be used to create/update GoodData objects directly under writer's credentials (see http://docs.keboolagooddatawriter.apiary.io/#proxy)

## 2013-11-26 (v 4.12)
- POST API call `project-invitations` merged into POST `project-users`
- POST API call `project-users` with `createUser` param will create user if missing in GD
- removed POST API call `project-invitations`
- new DELETE API call `project-users` for removing user from project and cancel his invitations

## 2013-22-20 (v 4.11)
- improved logging of GoodData tasks (accessible in `log` field of job status API call)

## 2013-11-19 (v 4.10)
- added API call `execute-reports`
- removed sanitization from upload jobs

## 2013-11-05 (v 4.9)
- decision if create/update dataset/dimension made on real situation in each GD project (lastExportDate is ignored but so far stays because of backwards compatibility with KBC UI)

## 2013-10-29 (v 4.8)
- added GET API call `proxy` with parameter `query` which can be used to query GoodData directly under writer's credentials

## 2013-10-29 (v 4.7)
- queue backend migrated to Amazon SQS
- possibility to use parameter `queue=secondary` to create a job in second queue which is served separately from primary queue used by default

## 2013-10-14 (v 4.6)
- filtered data load to cloned projects (see http://documentation.keboola.com/writers/gooddata#Project_cloning)

## 2013-09-23 (v 4.5)
- data load and date dimensions migrated from CL tool to Rest API

## 2013-07-29 (v 4.4)
- ExecuteReports job migrated from CL tool to Rest API

## 2013-07-25 (v 4.3)
- private links to S3

## 2013-07-04 (v 4.2)
- [UI] bugifxes
- [UI] rewritten to Angular
- tables export from SAPI using SAPI v2 with gzip
- addition of admin to project will send GD invitations to all its writers
- handling of GD bug with 401 error by exponential backoff

## 2013-05-28 (v 4.1)
- [API] added support for mandatory user filters, see <http://documentation.keboola.com/writers/gooddata/user-filters>

## 2013-05-20 (v 4.0)
- [API] added support for cloning GD projects and creating users, see <http://docs.keboolagooddatawriter4.apiary.io>
- [API] parameter `table` for `upload-table` changed to `tableId`
- [UI] creating and updating of dataset and loading data is hidden under summary `update-table` job
- [UI] `execute-reports` job is hidden from jobs queue
- [UI] added beta version of data model with visualization of references and date dimensions
- [UI] improvements and bug fixes in tables list and jobs queue

## 2013-03-13 (v 3.5)
- [UI] jobs queue speed improvements
- [UI] jobs queue rebuilt using Javascript
- [UI] dates and times are converted to user time zone
- upload of single table does not execute reports
- incremental load from SAPI using number of days in parameter `incrementalLoad`


## 2013-01-09 (v 3.4)
- creation of GoodData projects moved from Connection to writer - new api calls `/create-project` and `/invite-user`, see <http://docs.keboolagooddatawriter.apiary.io/#post-%2Fcreate-project>
- creation of each project creates new dedicated GoodData user whose credentials will be used for performing jobs (are stored in writer's configuration bucket)
- [UI] performance tweaks for loading configuration from SAPI
- [UI] forced uniqueness of names of datasets and date dimensions


## 2012-11-28 (v 3.3)
- [UI] added possibility to define GoodData access token in projects creation
- [UI] removal of writer (config is deleted instantly, GD project is scheduled for removal after one month)
- [UI] jobs queue UI integrated directly to Connection
- [UI] preview of uploaded csv in jobs queue (first 500 rows)
- [UI] preview of first ten values from csv in columns configuration
- updates for SAPI client v 2.5 compatibility
- added mandatory parameter `writerId` for `/job` and `/batch` observation calls
- added call `/jobs` allowing observation of jobs queue for given `writerId`
- removed deprecated API calls for direct communication with GoodData
- removed support for dedicated static auth token, every call needs Storage API token now
- added reports execution to the end of each batch load

## 2012-11-15 (v 3.2)
- added optional parameter `incrementalLoad` for `/upload-table` and `/upload-project` calls to force incremental or full load
- added optional parameter `sanitize` for `/load-data`, `/upload-table` and `/upload-project` calls to force cleaning or skipping null values
- added optional attributes `incrementalLoad` and `sanitize` for configuration tables
- solved dealing with recursion in tables references
- improved dealing with GoodData maintenance

Job/LoadData.php
<?php
/**
 * @author Jakub Matejka <jakub@keboola.com>
 * @date 2013-04-17
 */

namespace Keboola\GoodDataWriter\Job;

use Keboola\GoodDataWriter\Exception\WrongConfigurationException,
	Keboola\GoodDataWriter\Exception\JobProcessException,
	Keboola\GoodDataWriter\GoodData\CLToolApiErrorException;

class LoadData extends GenericJob
{
	public function run($job, $params)
	{
		if (empty($job['xmlFile'])) {
			throw new WrongConfigurationException("Parameter 'xmlFile' is missing");
		}
		$this->configuration->checkGoodDataSetup();

		$startTime = time();

		$xmlFile = $job['xmlFile'];
		if (!is_file($xmlFile)) {
			$xmlFilePath = tempnam($this->tmpDir, 'xml');
			exec('curl -s -L ' . escapeshellarg($xmlFile) . ' > ' . escapeshellarg($xmlFilePath));
			$xmlFile = $xmlFilePath;
		}

		libxml_use_internal_errors(TRUE);
		$xmlFileObject = simplexml_load_file($xmlFile);
		if (!$xmlFileObject) {
			$errors = '';
			foreach (libxml_get_errors() as $error) {
				$errors .= $error->message;
			}
			return $this->_prepareResult($job['id'], array(
				'status' => 'error',
				'error' => $errors,
				'debug' => $this->clToolApi->debugLogUrl
			), $this->clToolApi->output);
		}

		$projects = $this->configuration->getProjects();
		$gdJobs = array();

		// Create used date dimensions
		$dateDimensions = null;
		if ($xmlFileObject->columns) foreach ($xmlFileObject->columns->column as $column) if ((string)$column->ldmType == 'DATE') {
			if (!$dateDimensions) {
				$dateDimensions = $this->configuration->getDateDimensions();
			}

			$dimension = (string)$column->schemaReference;
			if (!isset($dateDimensions[$dimension])) {
				throw new WrongConfigurationException("Date dimension '$dimension' does not exist");
			}

			if (empty($dateDimensions[$dimension]['lastExportDate'])) {
				foreach ($projects as $project) if ($project['active']) {
					$gdJobs[] = array(
						'command' => 'createDate',
						'pid' => $project['pid'],
						'name' => $dateDimensions[$dimension]['name'],
						'includeTime' => !empty($dateDimensions[$dimension]['includeTime'])
					);
				}
				$this->configuration->setDateDimensionAttribute($dimension, 'lastExportDate', date('c', $startTime));
			}
		}

		$tableDefinition = $this->configuration->getTableDefinition($params['tableId']);
		$incrementalLoad = (isset($params['incrementalLoad'])) ? $params['incrementalLoad']
			: (!empty($tableDefinition['incrementalLoad']) ? $tableDefinition['incrementalLoad'] : 0);
		$sanitize = (isset($params['sanitize'])) ? $params['sanitize']
			: empty($tableDefinition['sanitize']);

		foreach ($projects as $project) if ($project['active']) {
			if (empty($tableDefinition['lastExportDate'])) {
				$gdJobs[] = array(
					'command' => 'createDataset',
					'pid' => $project['pid']
				);
			} else if (empty($tableDefinition['lastChangeDate']) || strtotime($tableDefinition['lastChangeDate']) > strtotime($tableDefinition['lastExportDate'])) {
				$gdJobs[] = array(
					'command' => 'updateDataset',
					'pid' => $project['pid']
				);
			}

			$gdJobs[] = array(
				'command' => 'loadData',
				'pid' => $project['pid']
			);
		}

		$sapiClient = new \Keboola\StorageApi\Client(
			$job['token'],
			$this->mainConfig['storageApi.url'],
			$this->mainConfig['user_agent']
		);
		$csvFile = $this->tmpDir . '/' . $job['id'] . '-' . uniqid() . '.csv';
		$options = array('format' => 'escaped');
		if ($incrementalLoad) {
			$options['changedSince'] = '-' . $incrementalLoad . ' days';
		}
		$sapiClient->exportTable($params['tableId'], $csvFile, $options);

		if ($sanitize) {
			$nullReplace = 'cat ' . escapeshellarg($csvFile) . ' | sed \'s/\"NULL\"/\"\"/g\' | awk -v OFS="\",\"" -F"\",\"" \'{';

			$i = 1;
			$columnsCount = $xmlFileObject->columns->column->count();
			foreach ($xmlFileObject->columns->column as $column) {
				$type = (string)$column->ldmType;
				$value = NULL;
				switch ($type) {
					case 'ATTRIBUTE':
						$value = '-- empty --';
						break;
					case 'LABEL':
					case 'FACT':
						$value = '0';
						break;
					case 'DATE':
						$format = (string)$column->format;
						$value = str_replace(
							array('yyyy', 'MM', 'dd', 'hh', 'HH', 'mm', 'ss', 'kk'),
							array('1900', '01', '01', '00', '00', '00', '00', '00'),
							$format);
						break;
				}
				if (!is_null($value)) {
					$testValue = '""';
					if ($i == 1) {
						$testValue = '"\""';
						$value = '\"' . $value;
					}
					if ($i == $columnsCount) {
						$testValue = '"\""';
						$value .= '\"';
					}
					$nullReplace .= 'if ($' . $i . ' == ' . $testValue . ') {$' . $i . ' = "' . $value . '"} ';
				}
				$i++;
			}
			$nullReplace .= '; print }\' > ' . escapeshellarg($csvFile . '.out');
			shell_exec($nullReplace);

			$csvFile .= '.out';
		}


		// Remove ignored columns from csv
		$ignoredColumns = array();
		$i = 1;
		foreach ($xmlFileObject->columns->column as $column) {
			if ((string)$column->ldmType == 'IGNORE') {
				$ignoredColumns[] = $i;
			}
			$i++;
		}
		if (count($ignoredColumns)) {
			shell_exec('cut -d"," -f' . implode(',', $ignoredColumns) . ' --complement ' . escapeshellarg($csvFile) . ' > '  . escapeshellarg($csvFile . '.i'));
			unlink($csvFile);
			rename($csvFile . '.i', $csvFile);
		}


		$gdWriteStartTime = date('c');
		$debug = array();
		$output = null;

		$datasetName = mb_strtolower(str_replace(' ', '', $xmlFileObject->name));
		$this->restApi->login($this->configuration->bucketInfo['gd']['username'], $this->configuration->bucketInfo['gd']['password']);

		$error = false;
		$this->clToolApi->setCredentials($this->configuration->bucketInfo['gd']['username'], $this->configuration->bucketInfo['gd']['password']);
		foreach ($gdJobs as $gdJob) {
			$this->clToolApi->debugLogUrl = null;
			try {
				switch ($gdJob['command']) {
					case 'createDate':
						$this->clToolApi->createDate($gdJob['pid'], $gdJob['name'], $gdJob['includeTime']);
						break;
					case 'createDataset':
						$this->clToolApi->createDataset($gdJob['pid'], $xmlFile);
						break;
					case 'updateDataset':
						$this->clToolApi->updateDataset($gdJob['pid'], $xmlFile, 1);

						break;
					case 'loadData':
						//$this->clToolApi->loadData($gdJob['pid'], $xmlFile, $csvFile, $incrementalLoad);

						// Get manifest
						$manifest = $this->restApi->get(sprintf('/gdc/md/%s/ldm/singleloadinterface/dataset.%s/manifest', $gdJob['pid'], $datasetName));
						$columnNames = array();
						foreach ($manifest['dataSetSLIManifest']['parts'] as &$column) {
							$columnNames[] = $column['columnName'];
							if ($incrementalLoad) {
								$column['mode'] = 'INCREMENTAL';
							}
						}

						// Add column headers according to manifest
						shell_exec('tail -n +2 ' . escapeshellarg($csvFile) . ' > ' . escapeshellarg($csvFile . '.1'));
						shell_exec('sed ' . escapeshellarg('1 i "' . implode('","', $columnNames) . '"') . ' ' . escapeshellarg($csvFile . '.1') . ' > ' . escapeshellarg($csvFile));



						die();

						break;
				}
			} catch (CLToolApiErrorException $e) {
				$this->clToolApi->output .= '!!! ERROR !!' . PHP_EOL . $e->getMessage() . PHP_EOL;
				$error = true;
			}

			if ($this->clToolApi->debugLogUrl) {
				$debug[$gdJob['command']] = $this->clToolApi->debugLogUrl;
			}
			$output .= $this->clToolApi->output;
		}

		if (empty($tableDefinition['lastExportDate'])) {
			$this->configuration->setTableAttribute($params['tableId'], 'export', 1);
		}
		$this->configuration->setTableAttribute($params['tableId'], 'lastExportDate', date('c', $startTime));
		return $this->_prepareResult($job['id'], array(
			'status' => $error ? 'error' : 'success',
			'debug' => json_encode($debug),
			'gdWriteStartTime' => $gdWriteStartTime,
			'gdWriteBytes' => filesize($csvFile),
			'csvFile' => $csvFile
		), $output);
	}
}
